craft_application/secrets.py

```diff
@@ -16,67 +16,103 @@
+
+    secret_strings: set[str]
+    """The actual secret text strings, to be passed to craft_cli."""
+
+
+def render_secrets(yaml_data: dict[str, Any], *, managed_mode: bool) -> BuildSecrets:
     """Render/expand the build secrets in a project's yaml data (in-place).
 
     This function will process directives of the form $(HOST_SECRET:<cmd>) in string
     values in ``yaml_data``. For each such directive, the <cmd> part is executed (with
-    bash) and the resulting output replaces the whole directive. The returned set
-    contains the outputs of all HOST_SECRET processing (for masking with craft-cli).
+    bash) and the resulting output replaces the whole directive. The returned object
+    contains the result of HOST_SECRET processing (for masking with craft-cli and
+    forwarding to managed instances).
 
     Note that only a few fields are currently supported:
 
     - "source" and "build-environment" for parts.
 
     Using HOST_SECRET directives in any other field is an error.
+
+    :param yaml_data: The project's loaded data
+    :param managed_mode: Whether the current application is running in managed mode.
     """
     command_cache: dict[str, str] = {}
 
+    if managed_mode:
+        command_cache = _decode_commands(os.environ)
+
     # Process the fields where we allow build secrets
     parts = yaml_data.get("parts", {})
     for part in parts.values():
-        _render_part_secrets(part, command_cache)
+        _render_part_secrets(part, command_cache, managed_mode)
 
     # Now loop over all the data to check for build secrets in disallowed fields
     _check_for_secrets(yaml_data)
 
-    return set(command_cache.values())
+    return BuildSecrets(
+        environment=_encode_commands(command_cache),
+        secret_strings=set(command_cache.values()),
+    )
 
 
 def _render_part_secrets(
-    part_data: dict[str, Any], command_cache: dict[str, Any]
+    part_data: dict[str, Any],
+    command_cache: dict[str, Any],
+    managed_mode: bool,  # noqa: FBT001 (boolean positional arg)
 ) -> None:
     # Render "source"
     source = part_data.get("source", "")
-    if (rendered := _render_secret(source, command_cache)) is not None:
+    if (rendered := _render_secret(source, command_cache, managed_mode)) is not None:
         part_data["source"] = rendered
 
     # Render "build-environment"
     build_env = part_data.get("build-environment", [])
     # "build-environment" is a list of dicts with a single item each
     for single_entry_dict in build_env:
         for var_name, var_value in single_entry_dict.items():
-            if (rendered := _render_secret(var_value, command_cache)) is not None:
+            rendered = _render_secret(var_value, command_cache, managed_mode)
+            if rendered is not None:
                 single_entry_dict[var_name] = rendered
 
 
-def _render_secret(text: str, command_cache: dict[str, str]) -> str | None:
+def _render_secret(
+    text: str,
```

Comment:
now that I see this, it is a strange varname, not asking for a change, as using "value" is just as generic.